implementations/micrometer-registry-new-relic/src/main/java/io/micrometer/newrelic/NewRelicMeterRegistry.java

```diff
@@ -15,82 +15,79 @@
      */
-    @SuppressWarnings("deprecation")
     public NewRelicMeterRegistry(NewRelicConfig config, Clock clock) {
-        this(config, clock, DEFAULT_THREAD_FACTORY,
-                new HttpUrlConnectionSender(config.connectTimeout(), config.readTimeout()));
+        //default to the HTTP/REST client
+        this(config, new NewRelicHttpClientProviderImpl(config), clock);
+    }
+    
+    /**
+     * @param config Configuration options for the registry that are describable as properties.
+     * @param clientProvider Provider of the HTTP or Agent-based client that publishes metrics to New Relic
+     * @param clock  The clock to use for timings.
+     */
+    public NewRelicMeterRegistry(NewRelicConfig config, NewRelicClientProvider clientProvider, Clock clock) {
+        this(config, clientProvider, new NewRelicNamingConvention(), clock, DEFAULT_THREAD_FACTORY);
     }
 
     /**
-     * @param config        Configuration options for the registry that are describable as properties.
-     * @param clock         The clock to use for timings.
+     * @param config Configuration options for the registry that are describable as properties.
+     * @param clientProvider Provider of the HTTP or Agent-based client that publishes metrics to New Relic
+     * @param namingConvention Naming convention to apply before metric publishing
+     * @param clock  The clock to use for timings.
      * @param threadFactory The thread factory to use to create the publishing thread.
      * @deprecated Use {@link #builder(NewRelicConfig)} instead.
      */
-    @Deprecated
-    public NewRelicMeterRegistry(NewRelicConfig config, Clock clock, ThreadFactory threadFactory) {
-        this(config, clock, threadFactory, new HttpUrlConnectionSender(config.connectTimeout(), config.readTimeout()));
-    }
-
-    // VisibleForTesting
-    NewRelicMeterRegistry(NewRelicConfig config, Clock clock, ThreadFactory threadFactory, HttpSender httpClient) {
+    @Deprecated  // VisibleForTesting
+    NewRelicMeterRegistry(NewRelicConfig config, NewRelicClientProvider clientProvider,  
+                NamingConvention namingConvention, Clock clock, ThreadFactory threadFactory) {
         super(config, clock);
 
-        if (!config.meterNameEventTypeEnabled() && StringUtils.isEmpty(config.eventType())) {
-            throw new MissingRequiredConfigurationException("eventType must be set to report metrics to New Relic");
-        }
-        if (StringUtils.isEmpty(config.accountId())) {
-            throw new MissingRequiredConfigurationException("accountId must be set to report metrics to New Relic");
+        if (clientProvider == null) {
+            throw new MissingRequiredConfigurationException("clientProvider required to report metrics to New Relic");
         }
-        if (StringUtils.isEmpty(config.apiKey())) {
-            throw new MissingRequiredConfigurationException("apiKey must be set to report metrics to New Relic");
+        if (namingConvention == null) {
```

Comment:
There's no way from a public constructor for `namingConvention` or `threadFactory` to be `null`. `MissingRequiredConfigurationException` is conventionally used for things users can configure that are required but might be missing.

Reply:
Ah, was being overly protective